test/extended/etcd/vertical_scaling.go

```diff
@@ -54,6 +55,64 @@ var _ = g.Describe("[sig-etcd][Feature:EtcdVerticalScaling][Suite:openshift/etcd
 		err = errors.Wrap(err, "pre-test: timed out waiting for initial cluster state to have 3 running machines and 3 voting members")
 		o.Expect(err).ToNot(o.HaveOccurred())
 
+		cpmsSupported, err := scalingtestinglibrary.PlatformSupportsCPMS(ctx, g.GinkgoT(), cpmsClient)
+		err = errors.Wrap(err, "pre-test: failed to determine if ControlPlaneMachineSet is present")
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		if cpmsSupported {
```

Comment:
do you think we can move that whole if code block into a dedicated test function? have the setup code above as an initializer for both

Reply:
I could move the setup code inside `BeforeEach()` so we do the common setup before each `It()` spec block.

And I would have two `It()` spec blocks in that case which would both always run except one would always get skipped based on the conditional. Just to illustrate what I'm thinking:
```Go

g.Describe("[sig-etcd][Feature:EtcdVerticalScaling][Suite:openshift/etcd/scaling] etcd [apigroup:config.openshift.io]", func() {

	var cpmsActive bool
	// Plus all the other common setup variables to pass to the It() closures below

	g.BeforeEach(func() {
		// Common setup and compute cpmsActive
		cpmsActive := ...
	})


	g.It("is able to vertically scale up and down with a single node when CPMS is active [Timeout:60m][apigroup:machine.openshift.io]", func() {
		if cpmsActive {
			Skip("CPMS is inactive so cannot use CPMS")
		}
		
		// New CPMS based workflow

	}


	g.It("is able to vertically scale up and down with a single node when CPMS is inactive [Timeout:60m][apigroup:machine.openshift.io]", func() {
		if cpmsActive {
			Skip("CPMS is active so cannot manually create machines")
		}
		
		// Old workflow
	}
	
}
```

Grammatically this is a bit awkward and not sure the conditional is too clear here.

I'm no ginkgo expert but I would think something like the following reads better with a `When` or `Describe` block which are both equivalent: 

```Go
g.When("CPMS is active", func() {

	g.It("is able to vertically scale up and down with a single node when CPMS is active [Timeout:60m][apigroup:machine.openshift.io]", func() {
	...
	}
}

g.When("CPMS is inactive", func() {

	g.It("is able to vertically scale up and down with a single node when CPMS is active [Timeout:60m][apigroup:machine.openshift.io]", func() {
	...
	}
}
```

But again I'm not sure how to conditionally run either `When()` block here so that's why I used `It()` in the former example.
But let me know if that's what you meant.

Reply:
I was thinking along the lines of 

> scalingtestinglibrary.SkipIfUnsupportedPlatform(ctx, oc)

so your first description fits the best. I just want to save some refactoring steps for [ETCD-330](https://issues.redhat.com/browse/ETCD-330) - where I would also need to detect whether CPMS active or not.

The g.When() indeed reads better, but if it's taking too much time for you to figure it out you can leave this for me :)

Reply:
Ah okay I get the context for this now. I'm still not liking the former layout of skipping one of the two `It()` blocks and haven't reviewed https://github.com/openshift/origin/pull/27461/files enough to see if we can't make that an entirely separate test case for that.

The quorum checker test would even need to disable the CPMS as well (when active) so we would have to serialize or order that so it can clean up and not be disruptive.

In the interest of landing a smaller fix temporarily (to improve pass rates and unblock other debugging efforts), I would lean towards punting the refactor for when we actually have a second test case. But not to say it's your problem :) I will refactor the test to make it more Ginkgo-readable as a follow up.

This would be easier to backport to 4.12 as well in case we have to refactor again in ETCD-330. WDYT?

Reply:
>  and haven't reviewed https://github.com/openshift/origin/pull/27461/files enough to see if we can't make that an entirely separate test case for that.

save yourself the time, I'll likely write this again from scratch. 

> The quorum checker test would even need to disable the CPMS as well (when active) so we would have to serialize or order that so it can clean up and not be disruptive.

I personally would just have the checker test run when there is no CPMS.
 
> I would lean towards punting the refactor for when we actually have a second test case. But not to say it's your problem :) I will refactor the test to make it more Ginkgo-readable as a follow up.

All good, I'm fine with refactoring this in the other ticket. Let's get the pass rates fixed first before we complicate.